src/azure-cli-core/azure/cli/core/azclierror.py

```diff
@@ -4,85 +4,178 @@
-        :param command: The command which brings the error.
-        :type command: str
-        :param recommendations: The recommendations to resolve the error.
-        :type recommendations: list
-        """
-        self.error_type = error_type
+    """ Base class for all the AzureCLI defined error classes. """
+
+    def __init__(self, error_msg):
+        # error message
         self.error_msg = error_msg
-        self.raw_exception = raw_exception
-        self.command = command
+        # set recommendations to fix the error if the message is not actionable,
+        # they will be printed to users after the error message, one recommendation per line
         self.recommendations = []
+        # exception trace for the error
+        self.exception_trace = None
         super().__init__(error_msg)
 
     def set_recommendation(self, recommendation):
         self.recommendations.append(recommendation)
 
-    def set_raw_exception(self, raw_exception):
-        self.raw_exception = raw_exception
+    def set_exception_trace(self, exception_trace):
+        self.exception_trace = exception_trace
 
     def print_error(self):
         from azure.cli.core.azlogging import CommandLoggerContext
         with CommandLoggerContext(logger):
-            message = '{}: {}'.format(self.error_type.value, self.error_msg)
+            # print error type and error message
+            message = '{}: {}'.format(self.__class__.__name__, self.error_msg)
             logger.error(message)
-            if self.raw_exception:
-                logger.exception(self.raw_exception)
+            # print exception trace if there is
+            if self.exception_trace:
+                logger.exception(self.exception_trace)
+            # print recommendations to action
             if self.recommendations:
                 for recommendation in self.recommendations:
                     print(recommendation, file=sys.stderr)
 
     def send_telemetry(self):
-        import azure.cli.core.telemetry as telemetry
-        telemetry.set_error_type(self.error_type.value)
-
-        # For userfaults
-        if self.error_type in [AzCLIErrorType.CommandNotFoundError,
-                               AzCLIErrorType.ArgumentParseError,
-                               AzCLIErrorType.ValidationError,
-                               AzCLIErrorType.ManualInterrupt]:
-            telemetry.set_user_fault(self.error_msg)
-
-        # For failures: service side error, client side error, unexpected error
-        else:
-            telemetry.set_failure(self.error_msg)
-
-        # For unexpected error
-        if self.raw_exception:
-            telemetry.set_exception(self.raw_exception, '')
+        telemetry.set_error_type(self.__class__.__name__)
+# endregion
+
+
+# region: Second Layer
+# Main categories of the AzureCLI error types, used for Telemetry alalysis
+# DO NOT raise the error classes here directly in your codes.
+class UserFault(AzCLIError):
```

Comment:
good point from @qwordy . how about `_UserFault` with `_` prefix to hint user not to use those?

Reply:
We can use Abstract Base Classes.
https://docs.python.org/3/library/abc.html
https://www.python.org/dev/peps/pep-3119/

Reply:
I don't it is revolved. Reopen the conversation.